app/scripts/background.js

```diff
@@ -176,6 +178,103 @@ const sendReadyMessageToTabs = async () => {
+ * domain that our phishing detection system is not aware of.
+ *
+ * @param {MetamaskController} theController
+ */
+function maybeDetectPhishing(theController) {
+  async function redirectTab(tabId, url) {
+    try {
+      return await browser.tabs.update(tabId, {
+        url,
+      });
+    } catch (error) {
+      return sentry?.captureException(error);
+    }
+  }
+  // we can use the blocking API in MV2, but not in MV3
+  const blocking = !isManifestV3;
+  browser.webRequest.onBeforeRequest.addListener(
+    (details) => {
+      if (details.tabId === browser.tabs.TAB_ID_NONE) {
+        return {};
+      }
+
+      const onboardState = theController.onboardingController.store.getState();
+      if (!onboardState.completedOnboarding) {
+        return {};
+      }
+
+      const prefState = theController.preferencesController.store.getState();
+      if (!prefState.usePhishDetect) {
+        return {};
+      }
+
+      // ignore requests that come from our phishing warning page
+      if (
+        details.initiator &&
+        // compare normalized URLs
+        new URL(details.initiator).toString() === phishingPageHref
+      ) {
+        return {};
+      }
+
+      const { hostname, href, searchParams } = new URL(details.url);
+      if (inTest) {
+        if (searchParams.has('IN_TEST_BYPASS_EARLY_PHISHING_DETECTION')) {
+          // this is a test page that needs to bypass early phishing detection
+          return {};
+        }
+      }
+
+      theController.phishingController.maybeUpdateState();
+      const phishingTestResponse =
+        theController.phishingController.test(hostname);
+      if (!phishingTestResponse?.result) {
+        return {};
+      }
+
+      theController.metaMetricsController.trackEvent({
+        // should we differentiate between background redirection and content script redirection?
+        event: MetaMetricsEventName.PhishingPageDisplayed,
+        category: MetaMetricsEventCategory.Phishing,
+        properties: {
+          url: hostname,
+        },
+      });
+      const querystring = new URLSearchParams({ hostname, href });
+      const redirectUrl = new URL(phishingPageHref);
+      redirectUrl.hash = querystring.toString();
+      const redirectHref = redirectUrl.toString();
+
+      // blocking is better than tab redirection, as blocking will prevent
+      // the browser from loading the page at all
+      if (blocking) {
+        if (details.type === 'sub_frame') {
+          // redirect the entire tab to the
+          // phishing warning page instead.
+          redirectTab(details.tabId, redirectHref);
+          // don't let the sub_frame load at all
+          return { cancel: true };
+        }
+        // redirect the whole tab, even if it's a sub_frame request
+        return { redirectUrl: redirectHref };
+      }
+      // redirect the whole tab, even if it's a sub_frame request
+      redirectTab(details.tabId, redirectHref);
+      return {};
```

Comment:
If main frame:

* mv2: immediately redirects to our Phishing Warning Page. no requests are made to the phishing server
* mv3: a redirect to our phishing warning page is initiated and completed asynchronously. Any redirects from the phishing page may be followed, but ultimately end up being overridden by the redirect to our Phishing Warning Page.

If sub frame (iframe):

* mv2: cancels the iframe's navigation. No requests are made to the phishing server. A redirect to our Phishing Warning Page is made asynchronously in the main frame (the tab).
* mv3: a redirect to our phishing warning page is initiated in the main frame (the tab) and completed asynchronously. Any redirects from the phishing page may still be followed, but ultimately end up being overridden by the redirect to our Phishing Warning Page.